hs/src/Reach/Eval/Core.hs

```diff
@@ -3915,7 +3921,61 @@ evalStmt = \case
           SLV_DLVar {} -> select_all de_val
           _ -> impossible "switch mvar"
       locAtf (srcloc_after_semi "switch" a sp) $ retSeqn fr ks_ne
-  (s@(JSThrow a _ _) : _) -> illegal a s "throw"
+  ((JSThrow a e _) : _) -> do
+    (dv_ty, dv) <- locAtf (srcloc_jsa "throw" a) (evalExpr e) >>= compileTypeOf . snd
+    curSt <- readSt id
+    exn_ref <- asks e_exn
+    exn <- liftIO $ readIORef exn_ref
+    -- Ensure we are throwing inside of a `try` block
+    unless (e_exn_in_throw exn) $ expect_ Err_Throw_No_Catch
+    -- Set or ensure the type that this `try` block is expected to catch
+    case e_exn_ty exn of
+      Nothing -> liftIO $ modifyIORef exn_ref (\ ex -> ex { e_exn_ty = Just dv_ty })
+      Just ty
+        | ty /= dv_ty -> expect_ $ Err_Try_Type_Mismatch ty dv_ty
+        | otherwise   -> return ()
+    saveLift =<< withAt (`DLS_Throw` dv)
+    liftIO $ modifyIORef exn_ref (\ ex -> ex { e_exn_st = Just curSt })
+    st <- asks e_st
```

Comment:
This should be a mergeSt because you have to be compatible with all previous throws. Try an example where you have an if where one branch introduces a participant and the other doesn't but both throw